init

kylemumma opened the pull request.

volokluev commented on snuba/web/delete_query.py:
```diff
@@ -38,11 +39,16 @@ class DeletesNotEnabledError(Exception):
     storage: WritableTableStorage,
     columns: Dict[str, list[Any]],
     attribution_info: Mapping[str, Any],
+    max_ongoing_mutations: int = 5,
```
this number should be a runtime config option

MeredithAnya replied:
but can still use 5 as the default for the config

volokluev commented on snuba/web/delete_query.py:
```diff
@@ -73,6 +84,33 @@ def delete_from_storage(
 
     results: dict[str, Result] = {}
     attr_info = _get_attribution_info(attribution_info)
+
+    # fail if too many mutations ongoing
+    cluster = storage.get_cluster()
+    if cluster.is_single_node():
+        query = f"""
+    SELECT count() as cnt
+    FROM system.mutations
+    WHERE table IN ({", ".join(map(repr, delete_settings.tables))}) AND is_done=0
+"""
+    else:
+        query = f"""
+SELECT max(cnt)
+FROM (
+    SELECT hostname() as host, count() as cnt
+    FROM clusterAllReplicas('{cluster.get_clickhouse_cluster_name()}', 'system', mutations)
+    WHERE table IN ({", ".join(map(repr, delete_settings.tables))}) AND is_done=0
+    GROUP BY host
+    )
+"""
```
Make this piece that determines the amount of mutations its own function. Mock it in a test. then you can test this functionality

onkar commented on snuba/web/delete_query.py:
```diff
@@ -73,6 +84,33 @@ def delete_from_storage(
+        .results[0][0]
+    )
+    if num_ongoing_mutations > max_ongoing_mutations:
+        raise TooManyOngoingMutationsError()
```
Should this exception be raised with a message saying how many mutations are ongoing? Possibly log the exception with relevant details like how many concurrent mutations are allowed etc.?

MeredithAnya commented on snuba/web/delete_query.py:
```diff
@@ -73,6 +84,33 @@ def delete_from_storage(
+FROM (
+    SELECT hostname() as host, count() as cnt
+    FROM clusterAllReplicas('{cluster.get_clickhouse_cluster_name()}', 'system', mutations)
+    WHERE table IN ({", ".join(map(repr, delete_settings.tables))}) AND is_done=0
```
This will combine the counts for the tables right? Can we group by table and check the max of the results? So the max of the max for the cluster query, and it would be the max of the single node

MeredithAnya replied:
but if there are two tables for a node it will combine them

```
SELECT count(), table
FROM system.mutations
WHERE table in ('outcomes_raw_local', 'outcomes_hourly_local')
GROUP BY table
```
<img width="1171" alt="Screenshot 2024-09-27 at 2 29 02 PM" src="https://github.com/user-attachments/assets/a764f2d0-b9cb-4b01-97ff-08bf507a2188">

If you don't group by table, the total count is 19

This is for one node. This won't be a problem for the search issues table, but I'm just saying if we ever need to delete from multiple tables it will combine the counts

kylemumma replied:
Oh I see what you mean now, fixed.

kylemumma commented:
* I made MAX_ONGOING_MUTATIONS_FOR_DELETE a runtime config option
* I added how many mutation are ongoing and the max allowed to the error message
* I added a test (it doesnt rly test the sql but it tests the other logic. I wouldnt know how to test the sql bc it would involve fake data in clickhouse)
